docs/tutorial/Tutorial.md

```diff
@@ -0,0 +1,501 @@
+   
+### Collector: Implement garbage collection
+
+1. Make a new file, called `gc_works`. 
+2. Add the following import statements:
+    ```rust
+    use super::global::MyGC;
+    use crate::plan::CopyContext;
+    use crate::policy::space::Space;
+    use crate::scheduler::gc_works::*;
+    use crate::util::alloc::{Allocator, BumpAllocator};
+    use crate::util::forwarding_word;
+    use crate::util::{Address, ObjectReference, OpaquePointer};
+    use crate::vm::VMBinding;
+    use crate::MMTK;
+    use std::marker::PhantomData;
+    use std::ops::{Deref, DerefMut};
+    ```
+3. Add a new structure, `MyGCCopyContext`, with the type parameter `VM: VMBinding`. It should have the fields `plan:&'static MyGC<VM>` and `mygc: BumpAllocator`.
+4. Create an implementation block - `impl<VM: VMBinding> CopyContext for MyGCCopyContext<VM>`.
+   1. Add a type alias for VMBinding (given to the class as `VM`): `type VM: VM`. 
+   2. Add the following skeleton functions (taken from `plan/global.rs`):
+       ```rust
+       fn new(mmtk: &'static MMTK<Self::VM>) -> Self { };
+       fn init(&mut self, tls: OpaquePointer) { };
+       fn prepare(&mut self) { };
+       fn release(&mut self) { };
+       fn alloc_copy(`init
+           &mut self,
+           original: ObjectReference,
+           bytes: usize,
+           align: usize,
+           offset: isize,
+           semantics: AllocationSemantics,
+       ) -> Address {
+       };
+       fn post_copy(
+           &mut self,
+           _obj: ObjectReference,
+           _tib: Address,
+           _bytes: usize,
+           _semantics: AllocationSemantics,
+       ) {
+       }
+       ```
+   3. To `new`, add an initialiser for the class:
+       ```rust
+       Self {
+             plan: &mmtk.plan,
+             mygc: BumpAllocator::new(OpaquePointer::UNINITIALIZED, None, &mmtk.plan),
+         }
+       ```
+   4. In `init`, set the `tls` variable in the held instance of `mygc` to the one passed to the function.
+   5. In `prepare`, rebind the allocator to the tospace.
+   6. Leave `release` with an empty body.
+   7. In `alloc`, call the allocator's `alloc` function. Above the function, use an inline attribute (`#[inline(always)]`) to tell the Rust compiler to always inline the function.
+   8. In `post_copy` add the following code. Also, add an inline (always) attribute.
+       ```rust
+       forwarding_word::clear_forwarding_bits::<VM>(obj);
+       ```
+5. Add a new public structure, `MyGCProcessEdges`, with the type parameter `<VM:VMBinding>`. It will hold an instance of `ProcessEdgesBase` and `PhantomData`, and implement the Default trait:
+    ```rust
+    #[derive(Default)]
+    pub struct MyGCProcessEdges<VM: VMBinding> {
+        base: ProcessEdgesBase<MyGCProcessEdges<VM>>,
+        phantom: PhantomData<VM>,
+    }
+    ```
+6. Add a new implementations block `impl<VM:VMBinding> ProcessEdgesWork for MyGCProcessEdges<VM>`.
+   1. Add a VM type alias (`type VM = VM`).
```

Comment:
`VM` is the associate type for `ProcessEdgesWork`. We set it to `VM` which is the type parameter of `MyGCProcessEdges`.